proposals/0046-optimistic-cluster-restart-automation.md

```diff
@@ -0,0 +1,321 @@
+400ms * 81000 = 9 hours, we assume most restart decisions to be made in 9 
+hours. If a validator restarts after 9 hours past the outage, it cannot join 
+the restart this way. If enough validators failed to restart within 9 hours, 
+then fallback to the manual, interactive cluster restart method.
+
+When a validator enters restart, it uses silent repair shred version to avoid
+interfering with those outside the restart. There is slight chance that 
+the silent repair shred version would collide with the shred version after the
+"silent repair phase", but even if this rare case occurred, we plan to flush 
+gossip on successful restart before entering normal validator operation.
+
+To be extra cautious, we will also filter out `LastVotedForkSlots` and
+`HeaviestFork` in gossip if a validator is not in "silent repair phase".
+
+`LastVotedForkSlots` message will be written into gossip and then be
+distributed to all others in restart. Meanwhile the validator can immediately
+enter next step.
+
+### 2. "silent repair phase": Repair ledgers up to the restart slot
+
+The main goal of this step is to repair all blocks which could potentially be
+optimistically confirmed.
+
+We need to prevent false negative at all costs, because we can't rollback an 
+optimistically confirmed block. However, false positive is okay. Because when 
+we select the heaviest fork in the next step, we should see all the potential 
+candidates for optimistically confirmed slots, there we can count the votes and
+remove some false positive cases.
+
+However, it's also overkill to repair every block presented by others. When
+`LastVotedForkSlots` messages are being received and aggregated, a validator
+can categorize blocks missing locally into 3 categories: ignored, must-have, 
+and unsure. Depending on the stakes of validators currently in restart, some
+slots with too few stake can be safely ignored, some have enough stake they
+should definitely be repaired, and the rest would be undecided pending more
+confirmations.
+
+Assume `RESTART_STAKE_THRESHOLD` is 80% and that 5% restarted validators can
+change their votes from what they voted before the restart due to mistakes or 
+malicious behavior.
+
+When only 5% validators are in restart, everything is in "unsure" category.
+
+When 67% validators are in restart, any slot with less than
+67% - 5% - (100-67%) = 29% is in "ignored" category, because even if all 
+validators join the restart, the slot will not get 67% stake. When this 
+threshold is less than 33%, we temporarily put all blocks with >33% stake into  
+"must-have" category to speed up repairing. Any slot with between 29% and 33% 
+stake is "unsure".
+
+When 80% validators are in restart, any slot with less than
+67% - 5% - (100-80%) = 42% is in "ignored" category, the rest is "must-have".
+
+From above examples, we can see the "must-have" threshold changes dynamically 
+depending on how many validators are in restart. The main benefit is that a
```

Comment:
Do we actually need a "must-have" threshold that changes dynamically, or can we just wait to hear from `RESTART_STAKE_THRESHOLD=80%` of the validators before doing any sort of repair?

Reply:
if we hear back from 80%, and we set the must-have threshold to 42% (not 47% because assuming 5% malicious), aren't we guaranteed we'll find all the optimistic blocks, regardless of what the remaining 20% respond if/when they join?

Reply:
> set the hard limit at 42%, and block X is at 41% when 80% reported numbers, but suddenly another validator restarted with 1% stake and voted for X,

If the block is at 41% when 80% reported, it cannot be optimistically confirmed because it has at most 61% votes.

> the point is whether the list to repair will keep growing

It won't keep growing if we just ignore everything after the first 80%. No matter who else joins, we ignore it

Reply:
I think my main concern here is that different validators might have different views of the world. The one main assumption is that everyone will make the same choice if they have same view of the world. If we hard code that ignore everything after the first 80% and only repair at 41%, then it's possible A and B have a different notion who made the first 80% so what's in their 41% is also different.

Of course, dynamic limit can't totally eliminate different validators having different views of the world, but I was hoping repair can start earlier and common blocks can be more consolidated as we get closer to 80%. Because we start repairing very early and discard more and more "not needed" blocks as we get closer to 80%, I was hoping the chance that many blocks are "on the edge" is smaller, but I don't have a strict math proof at this point.

Reply:
After chatting with Carl, we decided that in reality dynamic limit might not make much difference, so for the first version we will use hard limit to make the algorithm simple:
- the limit is set at 42% when 80% join the restart, different people might have different 80%, but in reality that might not cause any problem
- whenever some block gets 42% the repair can start, this way we can start repairs long before 80% are online

We also decided if we hear multiple RestartLastVotedForkSlots from the same person, then the new one is used for now. As long as less than 5% are changing their votes back and forth, we should be fine. Vote change will be logged as warning.

Changed now.